From 1ce34cc55aa9adcea62572fea0811695dc6bfaf1 Mon Sep 17 00:00:00 2001
From: Jeff Layton <jlayton@redhat.com>
Date: Thu, 17 Apr 2014 19:05:51 -0400
Subject: [net] svcrpc: fix gss-proxy NULL dereference in some error cases

Message-id: <1397761551-19434-1-git-send-email-jlayton@redhat.com>
Patchwork-id: 78963
O-Subject: [RHEL7.0 PATCH] [BZ#1085616] svcrpc: fix gss-proxy NULL dereference in some error cases
Bugzilla: 1085616
RH-Acked-by: Steve Dickson <SteveD@redhat.com>
RH-Acked-by: J. Bruce Fields <bfields@redhat.com>

From: "J. Bruce Fields" <bfields@redhat.com>

We depend on the xdr decoder to set this pointer, but if we error out
before we decode this piece it could be left NULL.

I think this is probably tough to hit without a buggy gss-proxy.

Reported-by: Andi Kleen <andi@firstfloor.org>
Cc: Simo Sorce <simo@redhat.com>
Signed-off-by: J. Bruce Fields <bfields@redhat.com>
Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1085616
Upstream: c66080ad0a9295eaf0246a46949619f92c75faaa
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/sunrpc/auth_gss/gss_rpc_upcall.c b/net/sunrpc/auth_gss/gss_rpc_upcall.c
index f1eb0d1..458f85e 100644
--- a/net/sunrpc/auth_gss/gss_rpc_upcall.c
+++ b/net/sunrpc/auth_gss/gss_rpc_upcall.c
@@ -298,7 +298,8 @@ int gssp_accept_sec_context_upcall(struct net *net,
  if (res.context_handle) {
   data->out_handle = rctxh.exported_context_token;
   data->mech_oid.len = rctxh.mech.len;
-  memcpy(data->mech_oid.data, rctxh.mech.data,
+  if (rctxh.mech.data)
+   memcpy(data->mech_oid.data, rctxh.mech.data,
       data->mech_oid.len);
   client_name = rctxh.src_name.display_name;
  }
-- 
1.7.1